Fix: PHP 8.1 pgsql change of return values

DanCld opened the pull request:
As per https://php.watch/versions/8.1/PgSQL-resource, PostgreSQL
extension has moved away from resource objects and is now
returning opaque classes (PgSQL\Connection, Result, Lob).

These classes no longer provide default conversions to (int), so
a different approach is required, while maintaining compatibility
to older PHP versions.

DanCld committed:
Fix: PHP 8.1 pgsql change of return values

As per https://php.watch/versions/8.1/PgSQL-resource, PostgreSQL
extension has moved away from resource objects and is now
returning opaque classes (PgSQL\Connection, Result, Lob).

These classes no longer provide default conversions to (int), so
a different approach is required, while maintaining compatibility
to older PHP versions.

schengawegga commented:
Sourcecode changes looks good to me. I haven´t got any postgre database at the moment, but if there is anybody who can approve the funcionality, i will merge this PR in the master branch.

DanCld committed:
Fixing a bug in tableInfo() method

My previous commit did not check for the right result structure
before raising an exception. While looking into this, I have found
a deprecation issue with Postgresql server versions >= 12 that
needed fixing as well (adsrc column was removed).

Context and recommended replacement: https://www.postgresql.org/docs/12/release-12.html#:~:text=obsolete%20pg_attrdef.-,adsrc,-column%20(Peter%20Eisentraut

rossnick commented:
My dev team looked at the code, and it's in line with what was found when we first encountered the problem with pear and PHP >= 8.1. So look's good enough for me !

MarkMaldaba commented on DB/pgsql.php:
```diff
@@ -1119,4 +1159,3 @@ function _checkManip($query)
  * End:
  */
 
-?>
```
Did you mean to remove this line?  This file should be consistent with the rest of the library.

MarkMaldaba commented on DB/pgsql.php:
```diff
@@ -286,7 +294,13 @@ function connect($dsn, $persistent = false)
+
+        if (function_exists('pg_version')) {
+            $pg_ver = pg_version($this->connection);
+	    $this->pg_major_server_version = intval($pg_ver['server'] ?? 0);
```
The null coalescing operator requires PHP 7 or above, so can't be used here.

MarkMaldaba commented on DB/pgsql.php:
```diff
@@ -1110,6 +1127,29 @@ function _checkManip($query)
+                    ? get_resource_id($result) 
+                    : (int)$result
+                : spl_object_id($result);
+     }
```
I suspect precedence rules mean this all works correctly, but it would be clearer if you used parentheses around the inner expression.

(It would be even clearer if you used `if`/`else` blocks, but whether or not that matters is a decision for the primary maintainer.)

DanCld committed:
PR change as per comments received

MarkMaldaba commented on DB/pgsql.php:
```diff
@@ -1119,4 +1159,4 @@ function _checkManip($query)
  */
 
-?>
+ ?>
```
Nitpick: Errant space added.

MarkMaldaba reviewed (commented):
Minor nitpick, but I'm otherwise happy with this from a code-review perspective.

I don't have any way of testing the code, and I don't feel I have the authority to approve even if I did, but from what I can see this looks good to me.

DanCld committed:
Errant space removed as per comment

MarkMaldaba reviewed (approved):
Approved from a code-review perspective (inspection only).

I do NOT have the ability to run and test this right now.

rossnick commented:
I tested it (not that I'm more of a sysadmin than a dev), and I no longer have warnings with simple selects.

schengawegga commented on DB/pgsql.php:
```diff
@@ -1110,6 +1127,29 @@ function _checkManip($query)
+                ? ( function_exists('get_resource_id')
+                    ? get_resource_id($result)
+                    : (int)$result
+                ) : spl_object_id($result);
```
Thanks @DanCld for your changes.
But for backwards compatibility, there is still a little change needed.
Both functions `spl_object_id()` and `get_resource_id()` are not compatible with PHP5.
Could you check first, if the function `spl_object_id()` exists, and if not returning the integer casted `$result` instead, as you did it with the `get_resource_id()` function, please?

Then i´m going to merge your request into the main branch.

DanCld replied:
I don't think that's needed. The decision tree for PHP5 is:

1. is_resource(...) will return true for PHP < 8.1
2. get_resource_id() is defined for PHP >= 8.0, so the _false_ path is being picked;
3. consequently, (int)$result is evaluated

For PHP == 8.0, is_resource(...) will return true and get_resource_id() is available, so this path is chosen.

For PHP >= 8.1, is_resource(...) will return false, while spl_object_id() is defined starting with PHP7, so no worries here.

Later edit: is_resource(...) returning false and spl_object_id() being unavailable sounds like undefined behaviour to me, as pg_query() has never historically returned on success some other data type that is not accounted for.

schengawegga replied:
@DanCld Sorry, you're right. I misunderstood that pgsql returns a resource before PHP8.

I agree and going to merge your PR later, when I'm at home.

Thank you for your great contribution.

schengawegga reviewed (changes requested):
Please make changes for backwards compatibility to PHP5.
Thank you.

DanCld committed:
Added a catch-all statement to _resultId(...)

This completes the decision tree related to handling pg_() internal functions return types and function availability based on PHP versions. The statement is not normally expected to be executed with any current or historical PHP version, but having it seems like a good idea, at least for completeness.

schengawegga commented on DB/pgsql.php:
```diff
@@ -1110,6 +1127,33 @@ function _checkManip($query)
+                    ? spl_object_id($result)
+                    // catch-all statement
+                    : (int)$result
+                )
```
@DanCld After a quick check on PHP5, there is a parsing error, because of missing semicolon in line 1155.
Could you add that, please?

DanCld committed:
Semicolon added as per comment
